src/language/lexer.rs

```diff
@@ -372,24 +372,44 @@ mod tests {
   fn test_number() {
     let mut lexer = Lexer::new("-9 0 9 -0 -9 1234 -1234 0.0 -0.0 1.0 -1.012 -0.101 1.0e2 -1.0E2 \
                                 0.00e-002 -0.00E+002 9.1E+0");
-    test_next_token(&mut lexer, Token::Int(0,2,    -9));
-    test_next_token(&mut lexer, Token::Int(3,4,     0));
-    test_next_token(&mut lexer, Token::Int(5,6,     9));
-    test_next_token(&mut lexer, Token::Int(7,9,    -0));
-    test_next_token(&mut lexer, Token::Int(10,12,  -9));
-    test_next_token(&mut lexer, Token::Int(13,17,   1234));
-    test_next_token(&mut lexer, Token::Int(18,23,  -1234));
-    test_next_token(&mut lexer, Token::Float(24,27, 0.0));
-    test_next_token(&mut lexer, Token::Float(28,32,-0.0));
-    test_next_token(&mut lexer, Token::Float(33,36, 1.0));
-    test_next_token(&mut lexer, Token::Float(37,43,-1.012));
-    test_next_token(&mut lexer, Token::Float(44,50,-0.101));
-    test_next_token(&mut lexer, Token::Float(51,56, 1.0e2));
-    test_next_token(&mut lexer, Token::Float(57,63,-1.0E2));
-    test_next_token(&mut lexer, Token::Float(64,73, 0.00e-002));
-    test_next_token(&mut lexer, Token::Float(74,84,-0.00E+002));
-    test_next_token(&mut lexer, Token::Float(85,91, 9.1E+0));
+    test_next_token(&mut lexer, Token::Int(0,2,    "-9"));
+    test_next_token(&mut lexer, Token::Int(3,4,    "0"));
+    test_next_token(&mut lexer, Token::Int(5,6,    "9"));
+    test_next_token(&mut lexer, Token::Int(7,9,    "-0"));
+    test_next_token(&mut lexer, Token::Int(10,12,  "-9"));
+    test_next_token(&mut lexer, Token::Int(13,17,  "1234"));
+    test_next_token(&mut lexer, Token::Int(18,23,  "-1234"));
+    test_next_token(&mut lexer, Token::Float(24,27,"0.0"));
+    test_next_token(&mut lexer, Token::Float(28,32,"-0.0"));
+    test_next_token(&mut lexer, Token::Float(33,36,"1.0"));
+    test_next_token(&mut lexer, Token::Float(37,43,"-1.012"));
+    test_next_token(&mut lexer, Token::Float(44,50,"-0.101"));
+    test_next_token(&mut lexer, Token::Float(51,56,"1.0e2"));
+    test_next_token(&mut lexer, Token::Float(57,63,"-1.0E2"));
+    test_next_token(&mut lexer, Token::Float(64,73,"0.00e-002"));
+    test_next_token(&mut lexer, Token::Float(74,84,"-0.00E+002"));
+    test_next_token(&mut lexer, Token::Float(85,91,"9.1E+0"));
     test_next_token(&mut lexer, Token::Eof(85,91));
     assert_eq!(None, lexer.next());
   }
+
+  #[test]
+  fn test_symbol() {
+    let mut lexer = Lexer::new("@ !  :$ =    [  {   (|] } )      ...");
+    test_next_token(&mut lexer, Token::At(0,0));
```

Comment:
I would expect I can use the location information to get back the string captured in token from lexer input. e.g. lexer.input[0..0] = "@"